File: doku.php
#!/usr/bin/php
<?php
# -*- coding: utf-8 -*-
# Setup VIM: ex: noet ts=2 sw=2 :
#
# PHP side Bridge of accessing DokuWiki functions from Python.
# See README for details.
#
# Author: Elan Ruusamäe <glen@pld-linux.org>
# Version: 1.0
#
# You should probably adjust path to DOKU_INC.

if ('cli' != php_sapi_name()) die();

//add to following define of 'DOKU_INC' to your "doku.local.php" file and adjust the path:
//define('DOKU_INC', '/home/caddy/wikifarm/dokuwiki/dokuwiki/');
//define('DOKU_INC', "d:/website/wwwroot/dokuwiki/" );
require_once './doku.local.php';

require_once DOKU_INC.'inc/init.php';
require_once DOKU_INC.'inc/common.php';
require_once DOKU_INC.'inc/cli.php';

# disable gzip regardless of config, then we don't have to compress when converting
$conf['compression'] = 0; //compress old revisions: (0: off) ('gz': gnuzip) ('bz2': bzip)

# override start page, as there's currently configured temporary frontpage
$conf['start'] = 'start'; //name of start page

function strip_dir($dir, $fn) {
	global $conf;
	return end(explode($dir.'/', $fn, 2));
}

$action = $argv[1];
$argPage = $argv[2];
//filext = $argv[3];

switch ($action) {
case 'cleanID':
	echo cleanID($argPage);
	break;
case 'wikiFN':
	if ($argc > 3 && $argv[3]) {
		echo strip_dir($conf['olddir'], wikiFN($argPage, $argv[3]));
	} else {
		echo strip_dir($conf['datadir'], wikiFN($argPage));
	}
	break;
case 'mediaFN':
	echo strip_dir($conf['mediadir'], mediaFN($argPage));
	break;
case 'metaFN':
	echo strip_dir($conf['metadir'], metaFN($argPage, $argv[3]));
	break;
case 'getNS':
	echo getNS($argPage);
	break;
case 'getId':
	echo getId();
	break;
default:
	die("Unknown knob: {$action}");
}

File: doku.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Setup VIM: ex: noet ts=2 sw=2 :
#
# Python side Bridge of accessing DokuWiki functions from Python.
# See README for details.
#
# Author: Elan Ruusamäe <glen@pld-linux.org>
# Version: 1.0

import sys
import subprocess

from MoinMoin import log
logging = log.getLogger(__name__)

class DokuWiki:
	def __init__(self):
		self.callcache = {}

	def __getattr__(self, method):
		def wrap(method):
			def wrapped(*args):
				return self.__call(method, *args)
			return wrapped
		return wrap(method)

	def __call(self, method, *args):
		args = list(args)
		uargs = []
		for arg in args:
			try:
				arg.decode('utf-8')
				#already UTF-8 ready
				uargs.append(arg)
			except UnicodeError:
				uargs.append(arg.encode('utf-8'))
		key = "%s:%s" % (method, ",".join(uargs))
		if not self.callcache.has_key(key):
			cmd = ['php', './doku.php', method ] + uargs
			res = subprocess.Popen(cmd, stdin = None, stdout = subprocess.PIPE, stderr = sys.stderr, close_fds = False).communicate()
			self.callcache[key] = unicode(res[0].decode('utf-8'))
			print "%s->%s" % (cmd, self.callcache[key])
		return self.callcache[key]
